include/picongpu/plugins/openPMD/restart/LoadSpecies.hpp

```diff
@@ -0,0 +1,239 @@
+                1,
+                MPI_UINT64_T,
+                fullParticlesInfo,
+                1,
+                MPI_UINT64_T,
+                gc.getCommunicator().getMPIComm() ) );
+
+            for( size_t i = 0; i < gc.getGlobalSize(); ++i )
+            {
+                /* this comparison is potentially harmful, since the order of
+                   ranks is not necessarily the same in subsequent MPI jobs. But
+                   due to the wrong sorting by rank in
+                   `openPMDCountParticles.hpp` while calculating the
+                   `myParticleOffset` we have to immitate that. */
+                if( i < gc.getGlobalRank() )
+                    particleOffset += fullParticlesInfo[ i ];
+                if( i == gc.getGlobalRank() )
+                    totalNumParticles = fullParticlesInfo[ i ];
+            }
+
+            log< picLog::INPUT_OUTPUT >(
+                "openPMD: Loading %1% particles from offset %2%" ) %
+                ( long long unsigned )totalNumParticles %
+                ( long long unsigned )particleOffset;
+
+            openPMDFrameType hostFrame;
+            log< picLog::INPUT_OUTPUT >(
+                "openPMD: malloc mapped memory: %1%" ) %
+                speciesName;
+            /*malloc mapped memory*/
+            meta::ForEach<
+                typename openPMDFrameType::ValueTypeSeq,
+                MallocMemory< bmpl::_1 > >
+                mallocMem;
+            mallocMem( hostFrame, totalNumParticles ); // TODO
+
+            log< picLog::INPUT_OUTPUT >(
+                "openPMD: get mapped memory device pointer: %1%" ) %
+                speciesName;
+            /*load device pointer of mapped memory*/
+            openPMDFrameType deviceFrame;
+            meta::ForEach<
+                typename openPMDFrameType::ValueTypeSeq,
+                GetDevicePtr< bmpl::_1 > >
+                getDevicePtr;
+            getDevicePtr( deviceFrame, hostFrame );
+
+            meta::ForEach<
+                typename openPMDFrameType::ValueTypeSeq,
+                LoadParticleAttributesFromOpenPMD< bmpl::_1 > >
+                loadAttributes;
+            loadAttributes(
+                params,
+                hostFrame,
+                particleSpecies,
+                particleOffset,
+                totalNumParticles );
+
+            if( totalNumParticles != 0 )
+            {
+                pmacc::particles::operations::splitIntoListOfFrames(
+                    *speciesTmp,
+                    deviceFrame,
+                    totalNumParticles,
+                    restartChunkSize,
+                    localDomain.offset,
+                    totalCellIdx_,
+                    *( params->cellDescription ),
+                    picLog::INPUT_OUTPUT() );
+
+                // TODO: skipping the deallocation if totalNumParticles == 0
+                // is ok?
+                /*free host memory*/
+                meta::ForEach<
```

Comment:
This can stay inside the if because we do not allocate memory when zero elements will be created. You could remove the comment above.